### PR TITLE
refactor: use EmailMessage class

### LINKED ISSUE #101: Use `email.message.Message` instead of rolling own (`RFC822Message`)
`pyproject-metadata` currently rolls his own serialization of the metadata. The specification says that the metadata serialization format is what is understood by the `email.parser` module when using the `compat32` policy. I think it would therefore be easier and safer to simply use the `email.message` module to serialize the metadata.

I started working on this but I was sidetracked by the fact that `email.message.Message` is now described as "legacy" and `email.message.EmailMessage` did not work for me when using the `compat32` policy. I see that @henryiii is putting some work in this package (thanks!!) thus I just wanted to contribute this thought.

### PR DESCRIPTION
**Pull request by henryiii**
Make `RFC822` a subclass of `EmailMessage` and remove all reliance on custom behaviors.

Close #101.

### CODE REVIEW

**Comment by henryiii** [comment]
Will need to see if we have anyone using this API, so we know if we need back compat.

**Comment by henryiii** [comment]
Here's the 30 usages I could find:

https://github.com/search?type=code&q=path%3A.py+%28%22import+pyproject_metadata%22+OR+%22from+pyproject_metadata%22%29+NOT+is%3Afork

(Quite a few are actually forks or copies that GitHub doesn't know are forks)

I think dropping `.headers` and `.body` is safe, not seeing usages of it, I think it's an internal detail.

File: pyproject_metadata/__init__.py [hunk 1]
```diff
@@ -114,37 +115,36 @@ class ConfigurationWarning(UserWarning):
...
-                for line in lines[1:]:
-                    text += ' ' * 8 + line + '\n'
-        text += '\n'
-        if self.body:
-            text += self.body
-        return text
-
-    def __bytes__(self) -> bytes:
-        return str(self).encode()
+        return bytes(self).decode()
```
Comment (henryiii) [inline pyproject_metadata/__init__.py:147]:
```suggestion
        return bytes(self).decode("utf-8")
```

File: pyproject_metadata/__init__.py [hunk 2]
```diff
@@ -113,46 +115,36 @@ class ConfigurationWarning(UserWarning):
...
+@dataclasses.dataclass
+class _SmartMessageSetter:
+    """
+    This provides a nice internal API for setting values in an RFC822Message to
+    reduce boilerplate.
 
-    def items(self) -> list[tuple[str, str]]:
-        return self._headers.copy()
+    If a value is None, do nothing.
+    If a value contains a newline, indent it (may produce a warning in the future).
```
Comment (dnicolodi) [inline pyproject_metadata/__init__.py:125]: Uh? Do we actually need this? Why? Does this even work with `EmailMesage`? I would not be surprised if it normalized whitespace in headers or encode it somehow. Anyhow, how do multi-line strings end up here?

Comment (henryiii) [reply pyproject_metadata/__init__.py:125]: It produces the same thing we produce now. It was added in #2, and pretty sure it mostly shows up with multiline summaries:

* https://github.com/pypa/pyproject-metadata/issues/30
* https://github.com/pypa/pyproject-metadata/pull/112

I think we could add a warning for now, and then not include the behavior if we can make backward incompatible changes (#140).

Comment (henryiii) [reply pyproject_metadata/__init__.py:125]: (Note that in #2, the extended mechanism is only for the description, which is now in the body instead, and the actual mechanism included pipe chars).

**Comment by henryiii** [comment]
pre-commit.ci autofix

File: pyproject_metadata/__init__.py (hunk 2 above)
Comment (henryiii) [reply pyproject_metadata/__init__.py:125]: (and in scikit-build-core I manually disallow multiline summaries)

Comment (henryiii) [reply pyproject_metadata/__init__.py:125]: And you can actually read these multiline strings with `email.parse`, I did check it, it keeps the indention. Without indentation, they don't work, of course.

Comment (henryiii) [reply pyproject_metadata/__init__.py:125]: I pulled out and merged the preparation steps, so you can see just the move by itself.

**Pull request merged**